communication/src/AsyncTlsConnection.cpp

```diff
@@ -0,0 +1,390 @@
+  while (!out_queue_.empty() &&
+         out_queue_.front().send_time + STALE_MESSAGE_TIMEOUT < std::chrono::steady_clock::now()) {
+    auto diff = std::chrono::steady_clock::now() - out_queue_.front().send_time;
+    LOG_WARN(logger_,
+             "Message queued for peer " << peer_id_.value() << " for "
+                                        << std::chrono::duration_cast<std::chrono::seconds>(diff).count()
+                                        << " seconds, with size: " << out_queue_.front().msg.size()
+                                        << " dropped. Message is stale: Exceeded threshold of "
+                                        << STALE_MESSAGE_TIMEOUT.count() << " seconds.");
+    queued_size_in_bytes_ -= out_queue_.front().msg.size();
+    out_queue_.pop_front();
+  }
+  if (out_queue_.empty()) {
+    return;
+  }
+
+  auto self = shared_from_this();
+  startWriteTimer();
+  boost::asio::async_write(
+      *socket_, boost::asio::buffer(out_queue_.front().msg), [this, self](const boost::system::error_code& ec, auto _) {
+        if (ec) {
+          if (ec == boost::asio::error::operation_aborted || disposed_) {
+            // The socket has already been cleaned up and any references are invalid. Just return.
+            return;
+          }
+          LOG_WARN(logger_,
+                   "Write failed to node " << peer_id_.value() << " for message with size "
+                                           << out_queue_.front().msg.size() << ": " << ec.message());
+          return dispose();
+        }
+        // The write succeeded.
+        write_timer_.cancel();
+        std::lock_guard<std::mutex> guard(write_lock_);
+        queued_size_in_bytes_ -= out_queue_.front().msg.size();
+        out_queue_.pop_front();
+        if (!out_queue_.empty()) {
+          write();
+        }
+      });
+}
+
+void AsyncTlsConnection::createSSLSocket(boost::asio::ip::tcp::socket&& socket) {
+  socket_ = std::make_unique<SSL_SOCKET>(io_service_, ssl_context_);
+  socket_->lowest_layer() = std::move(socket);
+}
+
+void AsyncTlsConnection::initClientSSLContext(NodeNum destination) {
+  auto self = std::weak_ptr(shared_from_this());
+  ssl_context_.set_verify_mode(boost::asio::ssl::verify_peer);
+
+  namespace fs = boost::filesystem;
+  fs::path path;
+  try {
+    path = fs::path(tlsTcpImpl_.config_.certificatesRootPath) / fs::path(std::to_string(tlsTcpImpl_.config_.selfId)) /
+           "client";
+  } catch (std::exception& e) {
+    LOG_FATAL(logger_, "Failed to construct filesystem path: " << e.what());
+    ConcordAssert(false);
+  }
+
+  boost::system::error_code ec;
+  ssl_context_.set_verify_callback(
+      [this, self, destination](auto preverified, auto& ctx) -> bool {
+        if (self.expired()) return false;
+        return verifyCertificateClient(preverified, ctx, destination);
+      },
+      ec);
+  if (ec) {
+    LOG_ERROR(logger_, "Unable to set client verify callback" << ec.message());
```

Comment:
I think it might be better to use `LOG_FATAL`, because we terminate the app.
I guess that `LOG_FATAL` might flush the log immediately(just speculation :) ).

Reply:
good catch!